File: benchmarks/ocaml/fannkuch.ml
(* Fannkuch redux benchmark *)

(* Reverse the first n elements of the array *)
let flip arr n =
  let result = Array.copy arr in
  for i = 0 to (n - 1) / 2 do
    let temp = result.(i) in
    result.(i) <- result.(n - i - 1);
    result.(n - i - 1) <- temp
  done;
  result

(* Count flips required to flip elements to get back to original order *)
let fannkuch n =
  let p = Array.init n (fun i -> i) in
  let perm = Array.make n 0 in
  let count = Array.make n 0 in
  let rec loop max_flips checksum sign perm_count =
    if perm_count >= 10000 then max_flips
    else begin
      (* Copy permutation to perm *)
      for i = 0 to n - 1 do
        perm.(i) <- p.(i) + 1
      done;

      let first = p.(0) in
      let max_flips', checksum' =
        if first <> 0 then begin
          (* Count flips *)
          Array.fill count 0 n 0;
          
          let rec count_flips perm' flips =
            if perm'.(0) = 1 then flips
            else begin
              let k = perm'.(0) - 1 in
              let perm'' = flip perm' k in
              count_flips perm'' (flips + 1)
            end
          in
          
          let flips = count_flips perm 0 in
          let max_flips' = max max_flips flips in
          let checksum' = checksum + sign * flips in
          (max_flips', checksum')
        end else (max_flips, checksum)
      in

      (* Generate next permutation *)
      let sign' = -sign in
      
      let rec find_j j =
        if j = n then None
        else if p.(j-1) < p.(j) then Some j
        else find_j (j+1)
      in
      
      match find_j 1 with
      | None -> max_flips'  (* We're done *)
      | Some j ->
          let perm_count' = perm_count + 1 in
          
          (* Flip first j elements *)
          for i = 0 to j / 2 - 1 do
            let idx1 = i in
            let idx2 = if i mod 2 = 0 then j - i else j - i - 1 in
            let temp = p.(idx1) in
            p.(idx1) <- p.(idx2);
            p.(idx2) <- temp
          done;
          
          (* Handle special cases *)
          if j < 2 then begin
            (* Find new j *)
            let rec find_next_j j' i =
              if i = n then j'
              else if p.(i-1) > p.(i) then find_next_j (i+1) (i+1)
              else find_next_j j' (i+1)
            in
            let j' = find_next_j 1 1 in
            
            (* Rotate elements *)
            for i = 0 to j' - 2 do
              let k = ref i in
              let temp = p.(i) in
              while !k < j' - 1 do
                incr k;
                p.(!k - 1) <- p.(!k)
              done;
              p.(j' - 1) <- temp
            done;
            
            loop max_flips' checksum' sign' perm_count'
          end else begin
            (* Regular case - rotate elements *)
            let j' = j - 1 in
            let first_j = p.(j') in
            for i = j' downto 1 do
              p.(i) <- p.(i-1)
            done;
            p.(0) <- first_j;
            
            loop max_flips' checksum' sign' perm_count'
          end
    end
  in
  loop 0 0 1 0

let () =
  let n = 10 in
  let start_time = Unix.gettimeofday () in
  
  let result = fannkuch n in
  
  Printf.printf "Fannkuch(%d): %d\n" n result;
  
  let elapsed = int_of_float ((Unix.gettimeofday () -. start_time) *. 1000.0) in
  Printf.printf "Time taken: %d ms\n" elapsed;
  
  (* Approximate memory usage *)
  let memory_usage = Gc.stat () in
  Printf.printf "Memory used: %d KB\n" (memory_usage.minor_words |> int_of_float |> (fun x -> x / 1024))

File: benchmarks/perl/fannkuch.pl
#!/usr/bin/env perl
# Fannkuch redux benchmark

use strict;
use warnings;
use Time::HiRes qw(gettimeofday tv_interval);

# Reverse the first n elements of the array
sub flip {
    my ($p, $n) = @_;
    for (my $i = 0; $i < $n/2; $i++) {
        my $temp = $p->[$i];
        $p->[$i] = $p->[$n-$i-1];
        $p->[$n-$i-1] = $temp;
    }
}

# Count flips required to flip elements to get back to original order
sub fannkuch {
    my ($n) = @_;
    my @p = (0..$n-1);
    my @perm = (0) x $n;
    my @count = (0) x $n;
    my $maxFlips = 0;
    my $checksum = 0;
    
    my $permCount = 0;
    my $sign = 1;
    
    while (1) {
        # Copy permutation to perm
        for (my $i = 0; $i < $n; $i++) {
            $perm[$i] = $p[$i] + 1;
        }
        
        my $first = $p[0];
        if ($first != 0) {
            # Count flips
            @count = (0) x $n;
            
            my $flips = 0;
            while ($perm[0] != 1) {
                my $k = $perm[0] - 1;
                flip(\@perm, $k);
                $flips++;
                $perm[0] = $k + 1;
            }
            
            if ($flips > $maxFlips) {
                $maxFlips = $flips;
            }
            
            $checksum += $sign * $flips;
        }
        
        # Generate next permutation
        $sign = -$sign;
        my $j = 1;
        while ($j < $n && $p[$j-1] >= $p[$j]) {
            $j++;
        }
        
        if ($j == $n) {
            last;
        }
        
        $permCount++;
        
        my $firstJ = $p[$j];
        for (my $i = 0; $i < $j; $i++) {
            if ($i % 2 == 0) {
                my $temp = $p[$i];
                $p[$i] = $p[$j-$i];
                $p[$j-$i] = $temp;
            } else {
                my $temp = $p[$i];
                $p[$i] = $p[$j-$i-1];
                $p[$j-$i-1] = $temp;
            }
        }
        
        if ($j < 2) {
            $j = 1;
            for (my $i = 1; $i < $n; $i++) {
                if ($p[$i-1] > $p[$i]) {
                    $j = $i + 1;
                }
            }
            for (my $i = 0; $i < $j-1; $i++) {
                my $k = $i;
                my $temp = $p[$i];
                while ($k < $j-1) {
                    $k++;
                    $p[$k-1] = $p[$k];
                }
                $p[$j-1] = $temp;
            }
        } else {
            $j--;
            $firstJ = $p[$j];
            for (my $i = $j; $i > 0; $i--) {
                $p[$i] = $p[$i-1];
            }
            $p[0] = $firstJ;
        }
        
        if ($permCount >= 10000) {
            last;
        }
    }
    
    return $maxFlips;
}

# Main
my $n = 10; # Default size
my $start_time = [gettimeofday];

my $result = fannkuch($n);

printf "Fannkuch(%d): %d\n", $n, $result;

my $elapsed = tv_interval($start_time) * 1000;
printf "Time taken: %.2f ms\n", $elapsed;

# Perl doesn't have a standard way to get memory usage
print "Memory monitoring not available for Perl implementation\n";